### PR TITLE
Reloadable Disablers

### PR DESCRIPTION
**Pull request by Surrealaser**
### About The Pull Request
You can now load or unload a disabler's power cell like a magazine, through use of a power cell holding cartridge. Further, infrastructure/support has been added to do the same for other energy weapons (though this PR does not expand this functionality to other weapons which will require supplemental balance changes). A power cell cartridge is used in order to limit the total charge available to a Sec Officer at one time; these are scarce and require valuable resources and research to produce in the protolathe. A Sec Officer and Warden starts with one spare, and one loaded in his disabler.

For balance reasons, by default per this PR, you can't load a power cell cartridge with a cell that has greater max capacity than 1000. Unloading a cartridge from a weapon is instant and it takes 4 seconds to load a cartridge into one. Loading can be interrupted, and you must be completely still to complete this; it is not viable to reload in combat unless you have cover (usually in the context of a firefight), or you've otherwise got someone/something stopping someone from jarring/pushing/pulling you to interrupt the unload/reload.

### Why It's Good For The Game
Eliminates the need for stowing multiple disablers, more realistic and intuitive, improves consistency with the stun baton as its power cell can be loaded/unloaded. Doesn't permit unlimited disabler spam in practice due to easily interruptible reloads mid-combat (if you want that, you'll have to stick with multiple disablers I guess).

### Changelog
🆑
add: Disablers can now have their power cells loaded and unloaded like a magazine with an interruptible delay, using power cell cartridges. Power capacity of cells that can be loaded into the disabler is capped at 1000 (default capacity).
add: Wardens and Sec Officers now spawn with one spare power cell cartridge.
/🆑

### CODE REVIEW

**Comment by carlarctg** [comment]
disabler spam is already enough of an issue without on-the-go reloads

**Comment by Surrealaser** [comment]
> disabler spam is already enough of an issue without on-the-go reloads

It isn't really viable to reload mid-combat unless you have someone running interference for you, or are engaged in a firefight where you can take cover. If a 1.5 second delay per unload/load (3 seconds total) is thought to be too short, it can be increased.

**Comment by Fikou** [comment]
we like snowflake caps
we like infinite disablers

**Comment by Fhaxaris** [comment]
Why not do this for every energy projectile weapon? Also instead of just taking out the cell you probably should have to use a screwdriver like how you have to with stun batons, defibs, and inducers.

**Comment by Surrealaser** [comment]
@Fhaxaris I felt it would be too controversial to make it standard for most energy weapons. I would leave that to people who want to court those arguments for subsequent PRs.

I could make it use a screwdriver, and it initially was required, though I thought that it was clumsy. However, I have no real objection to reinstating that as being necessary to extract the cell though.

**Comment by JJRcop** [comment]
Instead of directly putting cells in the disablers, they should have to be placed in a special magazine that holds 1 cell. This way officers will have to prepare some cells specifically for quick reloading, unless they want to put the disabler away, take the cell out of the mag, put a new cell in, then reload the charged cell's mag.

**Comment by ExcessiveUseOfCobblestone** [comment]
killing the balance around energy weapons 😢

**Comment by Surrealaser** [comment]
@ExcessiveUseOfCobblestone These changes only affect the disabler, keeps the max charge as it is, and can't really be done mid-combat (unless the other guy has guns and is generally keeping his distance). What problem do you have with this?

**Comment by Dawson1917** [comment]
Disablers already have an enormous magazine size that exceeds what's required for any intended use of them - slowing down & stunning 1 person at a time. On top of that, it's already trivial to duck into a sec checkpoint / wrench a recharger in maint and recharge energy-based guns in a few seconds flat. There absolutely doesn't need to be any easier ways to recharge disablers or other energy weapons, and it should probably be made more difficult if anything.
BTW you can easily just duck into maint and alt-click a fire extinguisher to give yourself >3 seconds to reload a cartridge without fear.

**Comment by ATH1909** [comment]
If you can corrupt a yellow slime power cell 4 times in a row, you can get a self-recharging power cell that has a low enough maxcharge value to be slotted into a disabler (there's a 34.39% chance that the cell will become rigged to explode upon next being used due to the corruptions, but you can fix that by injecting (or extracting) any reagent (aside from adding 5u of plasma, etc.) into the cell to un-rig it).

**Comment by Surrealaser** [comment]
@Dawson1917 Unless a fight is taking place by a sec station specifically I don't find that's really the case in practice (and if you do retreat to a checkpoint you are 100% losing whatever suspect you're dealing with). Though it's true you can drag around a charger, that's pretty silly, and generally isn't done unless you're expecting serious trouble (like say Nukies). As to the 3 second recharge, yes, you can duck into maintenance, but the problem is similar to dragging a charger: it takes time, if you're bumrushed you can't really do it (in fact it's worse than charger dragging in this case since you **will** be interrupted), and if you want to chase down someone you can't really do it (because your suspect will be gone in 3 seconds. 

As stated in the OP, the appeal is mainly in the case of a fire fight or perhaps prolonged mass battles where you can rotate out of a fight and nothing critical is lost if you do, not for standard situations.

**Comment by Surrealaser** [comment]
@ATH1909 That's an interesting point; I'm not sure how much of a problem something like this would be in practice, or if that kind of work should be hedged against, but it would be easy enough to preclude self-charging cells from being loaded.

**Comment by Ratvar** [comment]
Maybe you should remove stunbaton cell change instead of powercreeping everything else

**Comment by Surrealaser** [comment]
@Ratvar All removing cell replacement from batons would do is encourage people to carry more batons (which they already do because it's faster and more practical than swapping cells).

**Comment by Surrealaser** [comment]
@JJRcop It's an interesting idea, though what do you feel it achieves that the loading delay doesn't? Wouldn't this make existing concerns worse if you could quickload magazines into the disabler?

**Comment by Dawson1917** [comment]
>As stated in the OP, the appeal is mainly in the case of a fire fight

Disablers have 20 shots, enough to easily take down 3 people with middling aim. If you ever find yourself in a situation where you need to take down more than 3 people, you should be using your AOE sec gear like flashbangs or _working in a group with your other officers_. Worsening the situation is that disablers are already stackable because they take up a minuscule amount of inventory space. 

>or perhaps prolonged mass battles where you can rotate out of a fight and nothing critical is lost if you do

As said above. If you use up all your ammo you don't get to zip off for three seconds and come back with yet another 20 shots. This will benefit antags more than sec.

**Comment by Surrealaser** [comment]
> Disablers have 20 shots, enough to easily take down 3 people with middling aim. If you ever find yourself in a situation where you need to take down more than 3 people, you should be using your AOE sec gear like flashbangs or _working in a group with your other officers_. Worsening the situation is that disablers are already stackable because they take up a minuscule amount of inventory space.
 
Flashbangs are still easily better vs massed enemies. The point I was making is that this change doesn't dramatically alter how the disabler is actually used: in prolonged firefights, it performs better than the status quo disabler (while better, more powerful options still exist). Outside of that, you essentially have more convenient reloading between engagements; it is just not going to be practical in standard sec situations to reload in the middle of an encounter. I agree that disablers being stackable is problematic, and disabler hording is one thing this change aims to address, since it paves the way for making disablers too big to fit into a backpack, or otherwise occupy more space (this is out of scope of the PR though unfortunately, unless I were to generalize this as a disabler rebalance).

> As said above. If you use up all your ammo you don't get to zip off for three seconds and come back with yet another 20 shots. This will benefit antags more than sec.

I'm not sure if it'll benefit antags more than Sec overall, since Sec generally has more people which enables rotation that allows you to benefit from the reload in an ongoing fight versus an individual. Having said that, you generally don't want to be in a situation where you have to spend 3 seconds being completely still in most situations as Sec per my prior post. Further, the time penalty could be increased to as much as 3 seconds per load/unload to total 6 and it'd still achieve what I'm aiming for.

**Comment by MonsWTF** [comment]
Make the emptying of the disabler instant, while inserting a new cell takes 4 seconds. A single long timer is more risky than 2 short ones, plus once the officer chooses to reload, they're locked into it and have no room to change their mind.

Will the cell removal/insertion be audible to others in the area?

**Comment by Surrealaser** [comment]
@MonsWTF Not a bad idea and easy enough to do. Cell removal/insertion is audible on completion.

**Comment by Shadowflame909** [comment]
Don't we already have charge packs for the admin only ctf pulse rifle?

I bring that up to say infinite charged disablers are broken.

You can slot a hypercharged slime core for an infinite stun baton. A hypercharged slime cell disabler or even a bluespace cell disabler would be devastating

I'm just glad battery powered laser guns aren't being offered. That made sec borgs truly evil.


I bring this up to say that an ez infinite disabler is not balanced.

You can already get infinite stun batons with help from science. Imagine a bluespace battery disabler or ta

**Comment by Ty-the-Smonk** [comment]
2 words charged yellow

**Comment by wesoda25** [comment]
its a cool idea, but would not be fun gameplay wise.

**Comment by Tlaltecuhtli** [comment]
ruins the gimmick of energy weaponry, ruins balance free headcackes for little no interesting gameplay

**Comment by Surrealaser** [comment]
@Shadowflame909 As stated, I'm adding an update to prevent self-charging cells from being used.

@Ty-the-Smonk I'm not sure what you mean by that; it you're referring to yellow slime self-chargers, self-chargers are going to be forbidden from use with the disabler.

@wesoda25 Unfun in what capacity? Having to run all the way back to a charger isn't fun. I agree facing down a sec officer with unlimited disabler charge is bad, but you can't practically recharge in combat without either getting dunked by someone bumrushing you, or letting your perp get away.

@Tlaltecuhtli People keep saying it ruins balance, but they keep failing to specify exactly how given the limitations.

**Comment by Shadowflame909** [comment]
Wouldn't it be simpler to just drag around an unanchored, upgraded recharger for maximum power-game if you're gonna do it like that?

If you haven't already coded it, please donut because I fear it'll be feature bloat for things that can already be done

Albeit in a bit roundabout manner.

**Comment by Surrealaser** [comment]
@Shadowflame909 But who actually drags around rechargers except on blob, cult and nuke ops? Plus it's kind of a PITA versus just carrying around cells. Granted, recharger dragging is more mechanically powerful given that you can't really interrupt it with a simple shove or movement, and is effectively unlimited, power grid providing. Better still is just gun hoarding after ordering spares from Cargo (which is a problem that should be dealt with but is beyond the scope of this PR; not unlike cell loading being barely used for stun batons because hoarding more is better).

I mean whether or not you're prevented from loading in self-charge cells, recharger dragging is going to be a thing.

Remember that ultimately, what this feature does in practice is make the disabler a little better in prolonged firefights, and more convenient to recharge between combats/engagements.

**Comment by MonsWTF** [comment]
Oof, I forgot about Blob. 

I mean, in the circumstance that this change moves upward from Disablers to damage-based weapons (presumably it would?), it would be better to create said device that @JJRcop suggested above. Special clips that Sec has to prep in advance in order to reload low-power cell into the energy weapon should be enough of a trade-off to avoid making Rechargers redundant.

**Comment by Surrealaser** [comment]
@MonsWTF Chargers remain relevant because they provide unlimited power, and don't require interruptible do-afters/channels which lock you down.

I think my main problem with @JJRcop 's proposal is that while conceptually cool, it seems like it would actually heightens concerns with power creep, because it would effectively eliminate half the delay involved in a reload, assuming that by 'quickload' he means the reload is instantaneous.

**Comment by Fox-McCloud** [comment]
reloadable energy weapons is absolute cancer.

It was tried out over on Paradise; you could change out the cell of an energy gun.

It led to horrific abuse of situations and insane amount of spam.

👎

**Comment by Surrealaser** [comment]
@Fox-McCloud With the same limits and implementation?

**Comment by JJRcop** [comment]
@Surrealaser I mean the latter, same loading time. Sorry for the misunderstanding.

> If he meant you have to load cells into a magazine to have them be reloadable at all, and thus limits your reloads, that's more compelling, though you'd probably need no more than 2 reloads 99.99% of the time. Giving each sec one cell mag as standard with their disabler could be a good idea.

**Comment by Surrealaser** [comment]
@JJRcop Would you be able to prepare more of these cell mags in an auto/protolathe and such, or otherwise be able to make them?

**Comment by JJRcop** [comment]
I was thinking the sec vending machine but yeah some way to make more isn't a bad thing.

**Comment by MonsWTF** [comment]
Limited amt in the vending machine, additional clips stuck behind tech fab research?

**Comment by Surrealaser** [comment]
@JJRcop You have to be careful with accessibility to extra mags, because otherwise this is basically just like allowing cells to reload directly except with an extra step and some prep.

**Comment by AdamElTablawy** [comment]
still means blobs get fucked if it's for all energy weapons

**Comment by Surrealaser** [comment]
@AdamElTablawy I'm not proposing this be expanded to weapons other than the disabler; if others want to deal with that controversy they're welcome to.

**Comment by Fox-McCloud** [comment]
The only limitation differences here is a very short `do_after`

While we didn't have that limitation, I'm not sure it makes all that much of a difference--it's still going to lead to more spam and officers walking around with 5 cells in their packpacks.

**Comment by Surrealaser** [comment]
@Fox-McCloud 3 seconds of doing nothing and non-movement is pretty material in most cases as a Sec Officer as anyone who plays the role should be able to attest; you definitely won't be reloading this mid-chase, or combat. 

As stated before if this is felt to be too short, it can easily be increased further.

**Comment by necromanceranne** [comment]
I'll be honest, this seems relatively tame compared to the true cancer of "I have four disablers in my bag and I'm going to pull one after the other and dual-wield them to spam a hail of bullets to overwhelm you" that we have presently.

**Comment by carlarctg** [comment]
> If you can corrupt a yellow slime power cell 4 times in a row, you can get a self-recharging power cell that has a low enough maxcharge value to be slotted into a disabler (there's a 34.39% chance that the cell will become rigged to explode upon next being used due to the corruptions, but you can fix that by injecting (or extracting) any reagent (aside from adding 5u of plasma, etc.) into the cell to un-rig it).

what's 'corrupting'? first time i've heard of this

**Comment by Tlaltecuhtli** [comment]
> 
> 
> I'll be honest, this seems relatively tame compared to the true cancer of "I have four disablers in my bag and I'm going to pull one after the other and dual-wield them to spam a hail of bullets to overwhelm you" that we have presently.

yeah with this you need only 1 disabler and go print 5 cells

**Comment by necromanceranne** [comment]
I mean, with more guns you can actually use them mid combat. They're normal sized they're reasonably ergonomic if you account for it.

With a reload time that roots you in place I'll hardly call that viable for a fight.

**Comment by Ty-the-Smonk** [comment]
This sounds like snowflake code for cell exclusion

**Comment by Fox-McCloud** [comment]
>3 seconds of doing nothing and non-movement is pretty material in most cases as a Sec Officer as anyone who plays the role should be able to attest; you definitely won't be reloading this mid-chase, or combat.

That's not the point; in situations like nuke ops or what have you,  where you have defensive perimeter, 3 seconds is nothing.

It also completely changes the dynamic of "I emptied my disabler; I have to return to a specific spot to recharge it or drag around a charger with me 24/7" Both of which takes much *MUCH* longer than 3 seconds to do.

Pretending like this doesn't change the dynamic and fixating on "but in the middle of actual combat!" is missing the larger picture; even *IN* combat, 3 seconds can be small in certain contextual situations ("oh no, I ran out; oh well, he's downed, I can just swap cells; he'll still be down when I'm done).

This absolutely will change the dynamic of how security approaches and utilizes disablers, and not for the better, IMO.

**Comment by Surrealaser** [comment]
> yeah with this you need only 1 disabler and go print 5 cells

@Tlaltecuhtli Bruh, a bag of guns is way better because there's no pausing that forces you to be stationary while doing nothing; it's not comparable.

@Fox-McCloud Charger dragging is actually better than this, because you can still move and do things while your gun is charging. Bag of guns is better than this. And yes, it does change the dynamic of recharging your disabler; that's the point. The main difference is that it's more convenient to recharge it between engagements, because you absolutely are not doing it during engagements, save rare/atypical ones like a prolonged firefight.

Again, 3 seconds may be too short; I'm down for increasing it to as much as 6. There will definitely be a sub-10 second do-after duration of some kind that will make combat reloading utterly prohibitive.

Further, if the number of reloads is thought to be a problem the cell cartridge idea is a solution to that.

**Comment by Surrealaser** [comment]
@Ty-the-Smonk What do you mean by snowflake code? The code is standardized for all energy weapons. You can then determine cell compatibility on a per weapon basis per standard variables.

**Comment by Geikkamir** [comment]
> I'll be honest, this seems relatively tame compared to the true cancer of "I have four disablers in my bag and I'm going to pull one after the other and dual-wield them to spam a hail of bullets to overwhelm you" that we have presently.

Doesn't really seem to be mutually exclusive. If carrying around multiple disablers is still objectively better, the same powergamers that are carrying around 3 disablers in their bag now are going to keep doing it after the change.

**Comment by Shadowflame909** [comment]
Here's the tea

The recharger has 20 bullets and people carry 3 of them

We don't want them to carry 3 of them plus batreries to turn 60 bullets into 120

This will prolong firefights, and will replace the recharger with a more robust system of battery recharger swapping.

Not good.

**Comment by Anonmare** [comment]
I remember inducers charging guns and it was cancerous, even if it was only restricted to disablers would be hella annoying. Energy weapons are balanced around the fact they *can't* be reloaded, it's muscling in on ballistic's truf

**Comment by Shadowflame909** [comment]
Vice Versa for Ballistics as well Surrealaser

If we don't get electrical ammo generators for my ballistic cargo weapons, then them colliding on each others turf would be 

u n b a l a n c e d

**Comment by Surrealaser** [comment]
@Geikkamir 

Yes, I agree, powergamers will still continue to use Bag of Guns (BoG), because that's hands down superior to carrying and unloading/loading cells. Unfortunately it's an entirely separate issue well beyond the scope of this PR unless I make it a disabler balancing PR. If it were up to me guns wouldn't normally fit in backpacks and most containers that can hold them wouldn't be draggable at maximum speed.

@Shadowflame909 

#1: Even if we didn't limit the number of cells that could be reloaded via the earlier proposed cell cartridge system (where cartridges are scarce and expensive/hard to produce). Ballistics still have the advantage of instant reloading/unloading, which is huge in a fight.

#2: This is a change that doesn't apply to energy weapons in general so much as disablers specifically. If it were generalized to all energy weapons, secondary balance changes would be 100% required.

#3: With Bag of Guns, you'll generally just want more guns, not cells; cells only in slots where you can't stash guns, but honestly, with that number of disablers, you should have more than enough shots at that point to stun the entire station. As stated, BoG is a separate issue beyond the scope of this PR that should itself be addressed.

#4: You can make ballistic ammo in lathes, as you can make cells in lathes. True it costs more to do so, and it's generally harder to do so via research requisites, but cell cartridges would address this.

@Anonmare 

Inductors working on disablers means for the cost of one item slot, you can fully recharge, by default, a disabler 10 times. It's probably the closest equivalent, but it's still significantly more powerful than what is being proposed, especially when higher tech batteries are used with the inductors. Moreover, adding the cartridge element would further and stringently limit the amount of recharging you can do between recharger trips.

**Comment by PKPenguin321** [comment]
@JJRcop had a really good suggestion

**Comment by Surrealaser** [comment]
@PKPenguin321 I'll probably adopt it in some form sometime in the next couple of days. As I said, the supply of the cartridges would have to be stringently controlled, otherwise it's basically just cells with extra steps.

**Comment by SkeletalElite** [comment]
I like the idea that unload is instant but make the load longer. like at least 4 seconds. Maybe up to 6.

**Comment by anconfuzedrock** [comment]
The idea of this only working for disablers and with certain power cells is arbitrary and directly against the spirit of ss13. It only exists for balance reasons, but power cell replacement for disablers is in and of itself unbalanced, just LESS unbalanced than this working with any type of cell and for all kinds of weapons. This is dumb.

**Comment by Surrealaser** [comment]
@anconfuzedrock I completely disagree per the litany of earlier expressed arguments.

Also, replaceable batteries for all energy weapons could work, but would require more extensive balance changes that lie beyond the scope of this PR.

**Comment by Surrealaser** [comment]
@JJRcop Added the cell cartridges. They can be produced at a protolathe for 5000 plastic, 1000 glass and 1000 gold by default. Requires basic electromagnetics, weaponry and advanced power research to produce.

File: code/modules/projectiles/cell_cartridges/cell_cartridges.dm [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	///whether eject sound should vary
+	var/eject_sound_vary = TRUE
+
+/obj/item/cell_cartridge/Initialize()
+	. = ..()
+	if(cell_type && !cell)
+		cell = new cell_type(src)
+
+	if(!dead_cell)
+		cell.give(cell.maxcharge)
```
Comment (JJRcop) [inline code/modules/projectiles/cell_cartridges/cell_cartridges.dm]: Cells by default spawn with full charge, so this doesn't work. Better to remove the `dead_cell` var and change the cell_type of `/obj/item/cell_cartridge/dead` to `/obj/item/stock_parts/cell/empty`

Comment (Surrealaser) [reply code/modules/projectiles/cell_cartridges/cell_cartridges.dm]: Alright

**Comment by JJRcop** [comment]
Since you can already make power cells in the techfab, even the security one, maybe they should spawn without cells at all?

**Comment by Surrealaser** [comment]
@JJRcop Honestly I was thinking of just making the power cells unswappable by default.

**Comment by Surrealaser** [comment]
What's this hourglass bit about per Travis? I didn't touch any of those files whatsoever.

**Comment by optimumtact** [comment]
If you want to reload your weapons, take some ballistics.

Blurring the lines between energy weapons and ballistics isn't helpful, if anything I want them to feel and handle even more differently than they currently do.

For these reasons, and given there's no compelling reason for adding this, it's a no from me.

**Comment by Surrealaser** [comment]
@optimumtact okay, bag of guns it is then

**Comment by PKPenguin321** [comment]
As it has been, as it shall be. So sayeth orange man

**Pull request closed without merging**